swim/handlers_test.go

```diff
@@ -0,0 +1,205 @@
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package swim
+
+import (
+	"errors"
+	"testing"
+	"time"
+
+	"golang.org/x/net/context"
+
+	"github.com/stretchr/testify/suite"
+	"github.com/uber/ringpop-go/shared"
+	"github.com/uber/ringpop-go/swim/test/mocks"
+	"github.com/uber/tchannel-go"
+)
+
+type HandlerTestSuite struct {
+	suite.Suite
+
+	testNode *testNode
+	peers    []*testNode
+
+	cluster *swimCluster
+
+	ctx       tchannel.ContextWithHeaders
+	ctxCancel context.CancelFunc
+}
+
+func (s *HandlerTestSuite) SetupTest() {
+	// Create a test node, on its own. Tests can join this to the cluster for
+	// testing, if they want.
+	s.testNode = newChannelNode(s.T())
+	s.peers = append(s.peers, s.testNode)
+
+	// Create a cluster for testing. Join these guys to each other.
+	s.cluster = newSwimCluster(4)
+	s.cluster.Bootstrap()
+
+	s.ctx, s.ctxCancel = shared.NewTChannelContext(500 * time.Millisecond)
+}
+
+func (s *HandlerTestSuite) TearDownTest() {
+	s.ctxCancel()
+
+	// Destroy any nodes added to the peer list during testing
+	for _, node := range s.peers {
+		if node != nil {
+			node.Destroy()
+		}
+	}
+	if s.cluster != nil {
+		s.cluster.Destroy()
+	}
+}
+
+func (s *HandlerTestSuite) TestGossipStartHandler() {
+	s.testNode.node.gossip.Stop()
+	s.Require().True(s.testNode.node.gossip.Stopped())
+
+	_, err := s.testNode.node.gossipHandlerStart(s.ctx, &arg{})
+	s.NoError(err, "calling handler should not result in error")
+
+	s.False(s.testNode.node.gossip.Stopped())
+}
+
+func (s *HandlerTestSuite) TestGossipStopHandler() {
+	s.testNode.node.gossip.Start()
+	s.Require().False(s.testNode.node.gossip.Stopped())
+
+	_, err := s.testNode.node.gossipHandlerStop(s.ctx, &arg{})
+	s.NoError(err, "calling handler should not result in error")
+
+	s.True(s.testNode.node.gossip.Stopped())
+}
+
+func (s *HandlerTestSuite) TestToggleGossipHandler() {
+	s.Require().True(s.testNode.node.gossip.Stopped())
+
+	var err error
+
+	// Toggle gossip on
+	_, err = s.testNode.node.gossipHandler(s.ctx, &arg{})
+	s.NoError(err, "calling handler should not result in error")
+	s.False(s.testNode.node.gossip.Stopped())
+
+	// Toggle gossip off
+	_, err = s.testNode.node.gossipHandler(s.ctx, &arg{})
+	s.NoError(err, "calling handler should not result in error")
+	s.True(s.testNode.node.gossip.Stopped())
+}
+
+func (s *HandlerTestSuite) TestAdminLeaveJoinHandlers() {
+	// Join test node to cluster
+	s.cluster.Add(s.testNode.node)
+	s.Require().Equal(5, s.testNode.node.CountReachableMembers(), "expect a cluster of 5")
```

Comment:
I know that you didn't define this method, but I want to call out some inconsistencies in our codebase for future reference. In `node.go` we define a function named `CountReachableMembers()`. In `hashring.go`, we define `ServerCount()`. We should go with `Count` being the prefix or the suffix, but not both.